### DISCUSSION TITLE
Using KeyCloak via Docker with an external PostgreSQL database

### DISCUSSION THREAD

**Discussion by ZXRobotum** [keycloak/keycloak#14029] [Q&A]
### Description

Hello all,

I am struggling to create a Dockerfile for KeyCloak that will allow me to use my external database server and its port.

Does anyone here have any ideas, suggestions on how I can do this?

Or would it make more sense for me to build an image that only accepts requests for Postgres and then pass them on?

Or who has another idea?

The point is that I have a central database server and I can monitor, maintain and back it up.
best regards from Berlin

Z. Matthias

**Comment by naushadh** [comment]
You can use the official Keycloak docker image [jboss/keycloak](https://hub.docker.com/r/jboss/keycloak/), which supports configuring the PostgreSQL database (see environment variables section).

**Comment by alexted** [comment]
You have read [this section](https://www.keycloak.org/server/all-config#_database) of the documentation, haven't you?

**Comment by ZXRobotum** [comment]
Hey Naushadh,
thank you very much for your answer.

I tried this, but unfortunately it failed because the parameter "-e ...." (as an example) is not known by the programme. 
The description on the page: "https://hub.docker.com/r/jboss/keycloak/" therefore does not quite match the programme.
Do you have any idea how I can set this? Should I leave out the "-e ..." everywhere?

Best regards from Berlin

Z. Matthias

**Comment by ZXRobotum** [comment]
Hey Alexted,
Thank you very much for your answer.

I tried that too with all the possible parameters listed there. But unfortunately the JDBC cannot establish the connection. I also do not have a "URL" for the DB server where this programme can dock. I assume that this can only work if I set up a PostgreSQL instance directly on the DOCKER as described there.

However, I would like to address an external DB instance that is NOT on this host.

Do you have another idea?
Kind regards 


Z. Matthias

**Comment by naushadh** [comment]
Can you share what command(s) you've tried and what the output(s) were?

This should just work (replacing the `<PLACEHOLDERS>` with your own database's credentials)

```bash
$ docker run \
    -e KEYCLOAK_USER=<USERNAME> \
    -e KEYCLOAK_PASSWORD=<PASSWORD> \
    -e DB_VENDOR=postgres \
    -e DB_ADDR=<HOST NAME> \
    -e DB_PORT=<PORT NUM> \
    -e DB_USER=<DB USERNAME> \
    -e DB_PASSWORD=<DB PASSWORD> \
    -p 8080:8080
    jboss/keycloak
```

**Comment by ZXRobotum** [comment]
Hey,
well..... I tried it in this way:
docker run --name <keycloak> -p 8080:8080  -e KEYCLOAK_ADMIN=<user> -e KEYCLOAK_ADMIN_PASSWORD=<password> jboss/keycloak DB_VENDOR=postgres POSTGRES_PORT=<no> POSTGRES_DATABASE=<dbname> POSTGRES_USER=<dbuser> POSTGRES_PASSWORD=<dbpasswd>

The Different between your and my command-line is, that you write the name of image after all parameters and I between the parameters. I would like to try it and give you a feedback.

**Comment by ZXRobotum** [comment]
Hey,
I tried it now and I´ve got this Error - Message:
 FATAL [org.keycloak.services] (ServerService Thread Pool -- 57) Error during startup: java.lang.RuntimeException: Failed to connect to database

Which version of PostgreSQL is supported in this software & should I use the IP address rather than the hostname?
Best Regards

Z. Matthias

**Comment by bcmedeiros** [comment]
I'm also looking for a way to setup the (logical) PostgreSQL database for a docker compose setup. I have the PostgreSQL container running, but no nice way to configure the initial database creation anywhere.

**Reply by shawkins** [reply]
See the quickstart setup of postgres - https://raw.githubusercontent.com/keycloak/keycloak-quickstarts/refs/heads/main/kubernetes/keycloak.yaml the pg container takes a POSTGRES_DB environment variable.
